tests/dspy/test_save.py

```diff
@@ -0,0 +1,284 @@
+import json
+from unittest import mock
+
+import dspy
+import dspy.teleprompt
+import pytest
+
+import mlflow
+from mlflow.models import Model, ModelSignature
+from mlflow.types.schema import ColSpec, Schema
+
+from tests.helper_functions import (
+    _assert_pip_requirements,
+    _compare_logged_code_paths,
+    _mlflow_major_version_string,
+    expect_status_code,
+    pyfunc_serve_and_score_model,
+)
+
+
+class CoT(dspy.Module):
+    def __init__(self):
+        super().__init__()
+        self.prog = dspy.ChainOfThought("question -> answer")
+
+    def forward(self, question):
+        return self.prog(question=question)
+
+
+@pytest.fixture
+def cleanup_fixture():
+    yield
+
+    dspy.settings.configure(lm=None, rm=None)
+
+
+def test_basic_save(cleanup_fixture):
+    dspy_model = CoT()
+    dspy.settings.configure(lm=dspy.OpenAI(model="gpt-4o-mini", max_tokens=250))
+
+    with mlflow.start_run() as run:
+        mlflow.dspy.log_model(dspy_model, "model")
+
+    # Clear the lm setting to test the loading logic.
+    dspy.settings.configure(lm=None)
+
+    model_path = "model"
+    model_url = f"runs:/{run.info.run_id}/{model_path}"
+    loaded_model = mlflow.dspy.load_model(model_url)
+
+    # Check that the global settings is popped back.
+    assert dspy.settings.lm.kwargs["model"] == "gpt-4o-mini"
+    assert isinstance(loaded_model, CoT)
+
+
+def test_save_compiled_model(cleanup_fixture):
+    train_data = ["What is 2 + 2?", "What is 3 + 3?", "What is 4 + 4?", "What is 5 + 5?"]
+    train_label = ["4", "6", "8", "10"]
+    trainset = [
+        dspy.Example(question=q, answer=a).with_inputs("question")
+        for q, a in zip(train_data, train_label)
+    ]
+
+    def dummy_metric(program):
+        return 1.0
+
+    random_answers = ["4", "6", "8", "10"]
+    lm = dspy.utils.DummyLM(answers=random_answers)
+    dspy.settings.configure(lm=lm)
+
+    dspy_model = CoT()
+    optimizer = dspy.teleprompt.BootstrapFewShot(metric=dummy_metric)
+    optimized_cot = optimizer.compile(dspy_model, trainset=trainset)
+
+    with mlflow.start_run() as run:
+        mlflow.dspy.log_model(optimized_cot, "model")
+
+    # Clear the lm setting to test the loading logic.
+    dspy.settings.configure(lm=None)
```

Comment:
```suggestion
```

Reply:
So this line is to remove the global dspy settings and test if the global settings are correctly configured at loading time.

Reply:
got it :)